File: C_plus_plus/Partition_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *partition(ListNode *head, int x) {
        ListNode *less = NULL,*greater = NULL,*lessLast = NULL,*greaterLast = NULL , *p = head;
        while(p){
            if(p->val < x){
                if(!less)
                    less = p;
                else
                    lessLast->next = p;
                lessLast = p;
            }
            else{
                if(!greaterLast)
                    greater = p;
                else
                    greaterLast->next = p;
                greaterLast = p;
            }
            p = p->next;
        }
        if(less){
            lessLast->next = greater;
            head = less;
        }
        if(greater){
            greaterLast->next = NULL;
            if(!less)
                head = greater;
        }
        return head;
    }
};

File: java/Partition_List.java
// Just record the small head & tail and big head and tail
// One pass
// Time is O(n), space is O(1).

public class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode sHead = null, sPre = null, bHead = null, bPre = null, p = head;
        while (p != null) {
            if (p.val < x) {
                if (sHead == null)
                    sHead = p;
                else 
                    sPre.next = p;
                sPre = p;
            }
            else {
                if (bHead == null)
                    bHead = p;
                else
                    bPre.next = p;
                bPre = p;
            }
            p = p.next;
        }
        if (bPre != null) bPre.next = null;
        if (sHead != null)
            sPre.next = bHead;
        return sHead == null ? bHead : sHead;
    }
}
